python/GUIDE.rst

```diff
@@ -0,0 +1,314 @@
+define our handlers as coroutines and yield to IO operations:
+
+.. code-block:: python
+
+    @app.register(KeyValue)
+    @tornado.gen.coroutine
+    def Set(request, response, tchannel):
+        key = request.args.key
+        value = request.args.value
+
+        # Simulate some non-blocking IO work.
+        yield tornado.gen.sleep(1.0)
+
+        values[key] = value
+
+You have probably noticed that all of these handlers are passed ``response`` and
+`tchannel` objects, in addition to a ``request``. The ``response`` object is
+available for advanced use cases where it doesn't make sense to return one
+object as a response body -- for example, long-lived connections that gradually
+stream the response back to the caller.
+
+The `tchannel` object contains context about the current request (such as
+Zipkin tracing information) and should be used to make requests to other
+TChannel services. (Note that this API may change in the future.)
+
+
+---------
+Hyperbahn
+---------
+
+As mentioned earlier, our service is listening on an ephemeral port, so we are
+going to register it with the Hyperbahn routing mesh. Clients will use this
+Hyperbahn mesh to determine how to communicate with your service.
+
+Let's change our `run` method to advertise our service with a local Hyperbahn
+instance:
+
+.. code-block:: python
+
+    def run():
+        app.listen()
+        app.advertise(['localhost:23000'], 'keyvalue-server')
+        ioloop.IOLoop.current().start()
+
+The `advertise` method takes a seed list of Hyperbahn routers and the name of
+the service that clients will call into. After advertising, the Hyperbahn will
+connect to your process and establish peers for service-to-service
+communication.
+
+Consult the Hyperbahn documentation for instructions on how to start a process
+locally.
+
+
+---------
+Debugging
+---------
+
+Let's spin up the service and make a request to it through Hyperbahn. Python
+provides ``tcurl.py`` script, but we need to use the `Node
+version <https://github.com/uber/tcurl>`_) for now since it has Thrift support.
+
+.. code-block:: bash
+
+    $ python keyvalue/server.py &
+    $ node tcurl -p localhost:23000 -t ~/keyvalue/thrift service KeyValue::Set -3 '{"key": "hello", "value": "world"}'
+    $ node tcurl -p localhost:23000 -t ~/keyvalue/thrift service KeyValue::Get -3 '{"key": "hello"}'
+    $ node tcurl -p localhost:23000 -t ~/keyvalue/thrift service KeyValue::Get -3 '{"key": "hi"}'
+
+Your service can now be accessed from any language over Hyperbahn + TChannel!
+
+
+-------------
+Python Client
+-------------
+
+Let's make a client call from Python in ``keyvalue/client.py``:
+
+.. code-block:: python
+
+    from tornado import gen
+    from tornado import ioloop
+    from tchannel.thrift import client_for
+
+    from service import KeyValue
+
+
+    @gen.coroutine
+    def run():
+        app_name = 'keyvalue-client'
+
+        app = TChannel(app_name)
+        app.advertise(['localhost:23000'], app_name)
```

Comment:
Yeah I was thinking it might make sense to take the hyperbahn stuff out since it's not actually relevant to the outside world yet :)